pep-0712.rst

```diff
@@ -233,7 +246,27 @@ appear to be similar to this approach.
 This works well for fairly simple types, but leads to ambiguity in expected
 behavior for complex types such as generics. E.g. For ``tuple[int, ...]`` it is
 ambiguous if the converter is supposed to simply convert an iterable to a tuple,
-or if it is additionally supposed to convert each element type to ``int``.
+or if it is additionally supposed to convert each element type to ``int``. Or
+for ``int | None``, which isn't callable.
+
+Deducing the attribute type from the return type of the converter
+-----------------------------------------------------------------
+
+Another idea would be to allow the user to omit the attribute's type annotation
+if providing a ``field`` with a ``converter`` argument. Although this would
+reduce the common repetition this PEP introduces (e.g. ``x: str = field(converter=str)``),
+it isn't clear how to best support this while maintaining the current dataclass
+semantics (namely, that the attribute order is preserved for things like the
+synthesized ``__init__``, or ``dataclasses.fields``). This is because there isn't
+a way in Python (today) to get the annotation-only attributes interspersed with
+un-annotated attributes in the order they were defined.
```

Comment:
Maybe I'm wrong, but I'm not even sure it's possible for a metaclass to handle this correctly. The metaclass `__new__` would receive a namespace that contains `__annotations__` and the un-annotated fields separately:

```python
class MyType(type):
    def __new__(cls, name, bases, ns):
        print(ns)
        return super().__new__(cls, name, bases, ns)
        
        
class Foo(metaclass=MyType):
    a: int
    b = 0
    c: str

...
# {'__module__': '__main__', '__qualname__': 'Foo', '__annotations__': {'a': <class 'int'>, 'c': <class 'str'>}, 'b': 0}
```

I do think the current Python implementation of how to collect the class body into a namespace is the limiter here. Which means no amount of magic (save for using the AST or similar) can save us.

Reply:
You could do it with a `__prepare__` that returns a special namespace that keeps track of when things are added both to the namespace itself and to its `__annotations__`. (Not that I'm recommending that.)

Reply:
This does make me think we could do it _all_ fields were explicitly marked as `field`, so I'll address that too.

Reply:
@JelleZijlstra :see_no_evil: (I know none of us would go forth with this, but for my own edification) Although wouldn't that _techincally_ require that the Python implementation be assigning to `__annotations__` one at a time in the same order that it's also assigning to the namespace? Does Python-the-spec require that? (I suspect no :sweat_smile: )

Reply:
And, for posterity, oh God...

```python
class EldritchHorror(dict):
    def __init__(self):
        self.order = []

    def __setitem__(self, name, value):
        if name == "__annotations__":
            value = Frankentations(self.order)
        self.order.append(name)
        super().__setitem__(name, value)

class Frankentations(dict):
    def __init__(self, order):
        self.order = order

    def __setitem__(self, name, value):
        self.order.append(name)
        return super().__setitem__(name, value)

class MyType(type):
    def __new__(cls, name, bases, ns):
        print(ns.order)
        return super().__new__(cls, name, bases, ns)

    @classmethod
    def __prepare__(cls, name, bases):
        return EldritchHorror()

class Foo(metaclass=MyType):
    a: int
    b = 0
    c: str



```

results in:
`# ['__module__', '__qualname__', '__annotations__', 'a', 'b', 'c']`

:expressionless:

Reply:
I don't know if it's explicitly guaranteed, but it would be very weird if the order in which annotations are assigned doesn't match the textual order in which variables are assigned. I don't think relying on that order is risky.

Reply:
I was thinking if they were collected and assigned wholesale afterwards, but point taken.

I should tweak the wording regardless because it _is_ possible, just ugly.